src/allmydata/test/test_storage_https.py

```diff
@@ -0,0 +1,232 @@
+        """
+        expected_hash = b"JIj6ezHkdSBlHhrnezAgIC/mrVQHy4KAFyL+8ZNPGPM"
+        certificate_text = b"""\
+-----BEGIN CERTIFICATE-----
+MIIDWTCCAkECFCf+I+3oEhTfqt+6ruH4qQ4Wst1DMA0GCSqGSIb3DQEBCwUAMGkx
+CzAJBgNVBAYTAlpaMRAwDgYDVQQIDAdOb3doZXJlMRQwEgYDVQQHDAtFeGFtcGxl
+dG93bjEcMBoGA1UECgwTRGVmYXVsdCBDb21wYW55IEx0ZDEUMBIGA1UEAwwLZXhh
+bXBsZS5jb20wHhcNMjIwMzAyMTUyNTQ3WhcNMjMwMzAyMTUyNTQ3WjBpMQswCQYD
+VQQGEwJaWjEQMA4GA1UECAwHTm93aGVyZTEUMBIGA1UEBwwLRXhhbXBsZXRvd24x
+HDAaBgNVBAoME0RlZmF1bHQgQ29tcGFueSBMdGQxFDASBgNVBAMMC2V4YW1wbGUu
+Y29tMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAv9vqtA8Toy9D6xLG
+q41iUafSiAXnuirWxML2ct/LAcGJzATg6JctmJxxZQL7vkmaFFPBF6Y39bOGbbEC
+M2iQYn2Qemj5fl3IzKTnYLqzryGM0ZwwnNbPyetSe/sksAIYRLzn49d6l+AHR+Dj
+GyvoLzIyGUTn41MTDafMNtPgWx1i+65lFW3GHYpEmugu4bjeUPizNja2LrqwvwFu
+YXwmKxbIMdioCoRvDGX9SI3/euFstuR4rbOEUDxniYRF5g6reP8UMF30zJzF5j0k
+yDg8Z5b1XpKFNZAeyRYxcs9wJCqVlP6BLPDnvNVpMXodnWLeTK+r6YWvGadGVufk
+YNC1PwIDAQABMA0GCSqGSIb3DQEBCwUAA4IBAQByrhn78GSS3dJ0pJ6czmhMX5wH
++fauCtt1+Wbn+ctTodTycS+pfULO4gG7wRzhl8KNoOqLmWMjyA2A3mon8kdkD+0C
+i8McpoPaGS2wQcqC28Ud6kP9YO81YFyTl4nHVKQ0nmplT+eoLDTCIWMVxHHzxIgs
+2ybUluAc+THSjpGxB6kWSAJeg3N+f2OKr+07Yg9LiQ2b8y0eZarpiuuuXCzWeWrQ
+PudP0aniyq/gbPhxq0tYF628IBvhDAnr/2kqEmVF2TDr2Sm/Y3PDBuPY6MeIxjnr
+ox5zO3LrQmQw11OaIAs2/kviKAoKTFFxeyYcpS5RuKNDZfHQCXlLwt9bySxG
+-----END CERTIFICATE-----
+"""
+        certificate = x509.load_pem_x509_certificate(certificate_text)
+        self.assertEqual(get_spki_hash(certificate), expected_hash)
+
+
+def async_to_deferred(f):
+    """
+    Wrap an async function to return a Deferred instead.
+    """
+
+    @wraps(f)
+    def not_async(*args, **kwargs):
+        return Deferred.fromCoroutine(f(*args, **kwargs))
+
+    return not_async
+
+
+class PinningHTTPSValidation(AsyncTestCase):
+    """
+    Test client-side validation logic of HTTPS certificates that uses
+    Tahoe-LAFS's pinning-based scheme instead of the traditional certificate
+    authority scheme.
+
+    https://cryptography.io/en/latest/x509/tutorial/#creating-a-self-signed-certificate
+    """
+
+    def to_file(self, key_or_cert) -> str:
+        """
+        Write the given key or cert to a temporary file on disk, return the
+        path.
+        """
+        path = self.mktemp()
+        with open(path, "wb") as f:
+            if isinstance(key_or_cert, x509.Certificate):
+                data = key_or_cert.public_bytes(serialization.Encoding.PEM)
+            else:
+                data = key_or_cert.private_bytes(
+                    encoding=serialization.Encoding.PEM,
+                    format=serialization.PrivateFormat.TraditionalOpenSSL,
+                    encryption_algorithm=serialization.NoEncryption(),
+                )
+            f.write(data)
+        return path
+
+    def generate_private_key(self):
+        """Create a RSA private key."""
+        return rsa.generate_private_key(public_exponent=65537, key_size=2048)
```

Comment:
There's also `allmydata.crypto.rsa.create_signing_keypair` which does the same thing.

Reply:
Just going to leave this here, feels like test code is OK to be a bit duplicative because we might want to tweak it in various ways.